Mathlib/Algebra/Module/CharacterModule.lean

```diff
@@ -0,0 +1,189 @@
+  inferInstanceAs (AddCommGroup (A →+ _))
+
+section module
+
+variable [Module R A]  [Module R B]
+
+instance : Module R (CharacterModule A) :=
+  Module.compHom (A →+ _) (RingEquiv.toOpposite _ |>.toRingHom : R →+* Rᵈᵐᵃ)
+
+variable {R A B}
+
+@[simp] lemma smul_apply (c : CharacterModule A) (r : R) (a : A) : (r • c) a = c (r • a) := rfl
+
+/--
+Given an abelian group homomorphism `f : A → B`, then `f⋆(L) := L ∘ f` defines a linear map
+between `B⋆` and `A⋆`
+-/
+@[simps] def dual (f : A →ₗ[R] B) : CharacterModule B →ₗ[R] CharacterModule A where
+  toFun L := L.comp f.toAddMonoidHom
+  map_add' := by aesop
+  map_smul' r c := DFunLike.ext _ _ fun x ↦ by
+    simp only [RingHom.id_apply, smul_apply]
+    rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply, smul_apply]
+    erw [smul_apply, f.map_smul]
+    rfl
+
+lemma dual_surjective_of_injective (f : A →ₗ[R] B) (hf : Function.Injective f) :
+    Function.Surjective <| dual f := by
+  have : Fact ((0 : ℚ) < 1) := ⟨by norm_num⟩
+  have inst1 : Injective (AddCommGroupCat.of <| ULift.{max uA uB, 0} <| AddCircle (1 : ℚ)) :=
+    AddCommGroupCat.injective_of_divisible _
+
+  rintro (g : _ →+ _)
+  let g' : AddCommGroupCat.of (ULift.{max uA uB} A) ⟶
+      AddCommGroupCat.of (ULift.{max uA uB} (AddCircle (1 : ℚ))) :=
+    AddCommGroupCat.ofHom <| (ULift.moduleEquiv (R := ℤ)).symm.toAddMonoidHom.comp g |>.comp
+      (ULift.moduleEquiv (R := ℤ)).toAddMonoidHom
+  let f' : AddCommGroupCat.of (ULift.{max uA uB} A) ⟶ AddCommGroupCat.of (ULift.{max uA uB} B) :=
+    AddCommGroupCat.ofHom <| (ULift.moduleEquiv (R := ℤ)).symm.toAddMonoidHom.comp f |>.comp
+      (ULift.moduleEquiv (R := ℤ)).toAddMonoidHom
+  have inst2 : Mono f'
+  · rw [AddCommGroupCat.mono_iff_injective]
+    exact (ULift.moduleEquiv (R := ℤ)).symm.injective.comp hf |>.comp
+      (ULift.moduleEquiv (R := ℤ)).injective
+  let L := Injective.factorThru g' f'
+  refine ⟨(ULift.moduleEquiv (R := ℤ)).toAddMonoidHom.comp L |>.comp
+    (ULift.moduleEquiv (R := ℤ)).symm.toAddMonoidHom, AddMonoidHom.ext fun _ ↦
+      (ULift.ext_iff _ _).mp <| DFunLike.congr_fun (Injective.comp_factorThru g' f') _⟩
+
+/--
+Two isomorphic modules have isomorphic character modules.
+-/
+def congr (e : A ≃ₗ[R] B) : CharacterModule A ≃ₗ[R] CharacterModule B :=
+  LinearEquiv.ofLinear
+    (dual e.symm) (dual e)
+    (LinearMap.ext fun c ↦ DFunLike.ext _ _ fun a ↦ by
+      simp only [LinearMap.coe_comp, Function.comp_apply, dual_apply, LinearMap.id_coe, id_eq]
+      rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply]
+      erw [e.apply_symm_apply])
+    (LinearMap.ext fun c ↦ DFunLike.ext _ _ fun a ↦ by
+      simp only [LinearMap.coe_comp, Function.comp_apply, dual_apply, LinearMap.id_coe, id_eq]
+      rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply]
+      erw [e.symm_apply_apply])
+
+open TensorProduct
+
+/--
+Any linear map `L : A → B⋆` induces a character in `(A ⊗ B)⋆` by `a ⊗ b ↦ L a b`
+-/
+@[simps] noncomputable def curry :
```

Comment:
I think the names `curry` and `uncurry` should be swapped. Cf. `Function.curry`.

Reply:
done